locator/tablets.cc

```diff
@@ -798,15 +815,53 @@ class tablet_effective_replication_map : public effective_replication_map {
     }
 
 public:
+
     tablet_effective_replication_map(table_id table,
                                      replication_strategy_ptr rs,
                                      token_metadata_ptr tmptr,
-                                     size_t replication_factor)
+                                     size_t replication_factor,
+                                     datacenter_to_index_map&& datacenter_map,
+                                     tablet_to_datacenter_replication_factor_list_map&& rf_map)
             : effective_replication_map(std::move(rs), std::move(tmptr), replication_factor)
             , _table(table)
             , _sharder(*_tmptr, table)
+            , _datacenter_map(std::move(datacenter_map))
+            , _tablet_to_dc_replication_factor_map(std::move(rf_map))
     { }
 
+    [[nodiscard]] size_t get_replication_factor(const dht::token id, const seastar::sstring& datacenter) const override {
+        tablet_logger.trace("Get RF for token {}, dc {}", id, datacenter);
+        const auto tablet = get_tablet_map().get_tablet_id(id);
+        if (tablet.id >= _tablet_to_dc_replication_factor_map.size()) [[unlikely]] {
+            on_internal_error(tablet_logger,
+                    format("tablet id [{}] is out of range of replication factor map max: [{}].", tablet.id, _tablet_to_dc_replication_factor_map.size() - 1));
+        }
+
+        const auto& dc_rf_map = _tablet_to_dc_replication_factor_map[tablet.id];
+        const auto dc_it = _datacenter_map.find(datacenter);
+
+        if (dc_it == _datacenter_map.end()) [[unlikely]] {
+            on_internal_error(tablet_logger, format("Data center: [{}] does not exist in the datacenter map.", datacenter));
+        }
+        if (dc_it->second >= dc_rf_map.size()) [[unlikely]] {
+            on_internal_error(tablet_logger, format("Data center's: [{}] index: [{}] is higher than the number of data centers in the dc_rf_map: [{}].",
+                                                     datacenter, dc_it->second, dc_rf_map.size()));
```

Comment:
Message says "index is higher than the number", which is not quite true, since we fail also when index is equal to the number of dcs. Accurate wording would be something like "index [] is out of range, max=[]"

Reply:
fixed